lib/rack/request.rb

```diff
@@ -232,52 +278,87 @@ def xhr?
         get_header("HTTP_X_REQUESTED_WITH") == "XMLHttpRequest"
       end
 
-      def host_with_port
-        port = self.port
-        if port.nil? || port == DEFAULT_PORTS[scheme]
-          host
+      # The `HTTP_HOST` header.
+      def host_authority
+        get_header(HTTP_HOST)
+      end
+
+      def host_with_port(authority = self.authority)
+        host, address, port = split_authority(authority)
+
+        if port == DEFAULT_PORTS[self.scheme]
+          return host
         else
-          host = self.host
-          # If host is IPv6
-          host = "[#{host}]" if host.include?(':')
-          "#{host}:#{port}"
+          return authority
         end
       end
 
+      # Returns a formatted host, suitable for being used in a URI.
       def host
-        # Remove port number.
-        strip_port hostname.to_s
+        host, address, port = split_authority(self.authority)
+
+        return host
+      end
+
+      # Returns an address suitable for being used with `getaddrinfo`.
+      def hostname
+        host, address, port = split_authority(self.authority)
+
+        return address
       end
 
       def port
-        result =
-          if port = extract_port(hostname)
-            port
-          elsif port = get_header(HTTP_X_FORWARDED_PORT)
-            port
-          elsif has_header?(HTTP_X_FORWARDED_HOST)
-            DEFAULT_PORTS[scheme]
-          elsif has_header?(HTTP_X_FORWARDED_PROTO)
-            DEFAULT_PORTS[extract_proto_header(get_header(HTTP_X_FORWARDED_PROTO))]
-          else
-            get_header(SERVER_PORT)
+        if authority = self.authority
+          host, address, port = split_authority(self.authority)
+          if port
+            return port
           end
+        end
+
+        if forwarded_port = self.forwarded_port
+          return forwarded_port.first
+        end
 
-        result.to_i unless result.to_s.empty?
+        if scheme = self.scheme
+          if port = DEFAULT_PORTS[self.scheme]
```

Comment:
If you are going to set the `scheme` local variable, you should use it as the argument to `DEFAULT_PORTS.[]`.  Otherwise don't define the local variable.